Execute Kubernetes tests in parallel

cbandy opened the pull request:
The crux of the solution is the counter in `setupKubernetes`. When a test calls it then `t.Parallel`, it shares the `envtest.Environment` with other tests that have done the same. When all such tests finish, the counter decrements to zero, and the environment is cleaned up.

Normally, `go test` allows [GOMAXPROCS parallel tests to run simulatenously](https://pkg.go.dev/cmd/go/internal/test). Some of our Kubernetes tests require more resources than others, so another function `require.ParallelCapacity` governs how many run at a time.

The idea is to let each test describe its "size" so that more "smaller" tests run together compared to fewer "bigger" tests. Something like,

```go
// ConfigMaps, Secrets, etc.
require.ParallelCapacity(t, 0)

// One replica.
require.ParallelCapacity(t, 1)

// Two replicas, or
// One replica plus one PgBouncer, or
// One replica plus one pgAdmin, etc.
require.ParallelCapacity(t, 2)

// Three replicas, etc.
require.ParallelCapacity(t, 3)
```

The current implementation assumes a capacity of one so that anything other than `0` runs alone. I imagine, in the future, the capacity should be configurable so the person running the suite can dial up the parallelism when they know the Kubernetes cluster can handle it.

As it stands, the time it takes to run the suite is reduced significantly.

On my under-powered Chromebook, `make check-envtest` takes around ten minutes and I must raise the test timeout. This PR takes less than two minutes.

<details>
  <summary>Before: <code>10m12.720s</code></summary>

```
● time make check-envtest GO_TEST='go test -timeout=15m'
KUBEBUILDER_ASSETS="…/hack/tools/envtest/bin" PGO_NAMESPACE="postgres-operator" go test -timeout=15m -count=1 -cover -tags=envtest ./...
…
ok      github.com/crunchydata/postgres-operator/internal/controller/postgrescluster    596.626s        coverage: 79.7% of statements
…
ok      github.com/crunchydata/postgres-operator/internal/upgradecheck  8.134s  coverage: 89.1% of statements
…

real    10m12.720s
user    10m23.519s
sys     2m54.173s
```

</details>

<details>
  <summary>After: <code>1m42.354s</code></summary>

```
● time make check-envtest
KUBEBUILDER_ASSETS="…/hack/tools/envtest/bin" PGO_NAMESPACE="postgres-operator" go test -count=1 -cover -tags=envtest ./...
…
ok      github.com/crunchydata/postgres-operator/internal/controller/postgrescluster    94.946s coverage: 79.7% of statements
…
ok      github.com/crunchydata/postgres-operator/internal/upgradecheck  8.157s  coverage: 89.1% of statements
…

real    1m42.354s
user    1m46.585s
sys     0m48.762s
```

</details>


<br/>

On my k3s VM, `make check-envtest-existing` also takes almost ten minutes. This PR takes six minutes.

<details>
  <summary>Before: <code>9m23.969s</code></summary>

```
● time make check-envtest-existing GO_TEST='go test -timeout=15m'
…
USE_EXISTING_CLUSTER=true PGO_NAMESPACE="postgres-operator" go test -timeout=15m -count=1 -cover -p=1 -tags=envtest ./...
…
ok      github.com/crunchydata/postgres-operator/internal/controller/postgrescluster    519.946s        coverage: 83.8% of statements
…
ok      github.com/crunchydata/postgres-operator/internal/upgradecheck  14.516s coverage: 96.0% of statements
…

real    9m23.969s
user    1m12.260s
sys     0m10.612s
```

</details>


<details>
  <summary>After: <code>6m3.493s</code></summary>

```
● time make check-envtest-existing

USE_EXISTING_CLUSTER=true PGO_NAMESPACE="postgres-operator" go test -count=1 -cover -p=1 -tags=envtest ./...
…
ok      github.com/crunchydata/postgres-operator/internal/controller/postgrescluster    317.367s        coverage: 84.2% of statements
…
ok      github.com/crunchydata/postgres-operator/internal/upgradecheck  15.235s coverage: 96.0% of statements
…

real    6m3.493s
user    0m36.161s
sys     0m7.087s
```

</details>

---

**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?
   - [x] Have you added automated tests?

**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [x] Testing enhancement

**What is the current behavior (link to any open issues here)?**

`envtest` tests run sequentially.

**What is the new behavior (if this is a feature change)?**

`envtest` tests can run in parallel. Most do.

**Other Information**:

Issue: [sc-12206]
Issue: [sc-12208]

jmckulk reviewed (commented):
I like it! To be clear, now it is the test writer's responsibility to consider how many pods will be created as part of the test. Then request resources accordingly

cbandy commented on internal/controller/postgrescluster/apply_test.go:
```diff
@@ -33,35 +33,20 @@ import (
-	cc, err := client.New(config, client.Options{})
-	assert.NilError(t, err)
+	env, cc := setupKubernetes(t)
+	t.Parallel()
```
> Would this be `require.ParallelCapacity(t, 0)`?

Yeah.

> Should we use both `t.Parallel` and `require.ParallelCapacity`?

We should probably use just one, right? That would be ParallelCapacity.
 
> I don't see any other `t.Parallel` so I think this could be a 🔧

I started with `require.ParallelCapacity` then noticed it was the first import of our own packages in this file. This file wasn't even using `setupEnvTest`... to make it clear that this was straight-Kubernetes? I dunno.

I'll change it to `require.ParallelCapacity(t, 0)` unless someone can think of a reason to avoid it.

cbandy commented on internal/controller/postgrescluster/instance_test.go:
```diff
@@ -1712,26 +1703,17 @@ func TestFindAvailableInstanceNames(t *testing.T) {
-	t.Cleanup(func() { teardownManager(cancel, t) })
+	ctx := context.Background()
+	_, tClient := setupKubernetes(t)
+	require.ParallelCapacity(t, 99)
```
> With capacity `99`, are we just assuming the upgrade test will be a tremendous job?

This is one of the tests I couldn't grok. It creates clusters without cleaning them up, some tests skip, and some things are mocked.

I can tell it's not zero, but I can't tell what number it should have.

> Maybe add a comment for cases where the capacity doesn't exactly match the number of resources that need to be created?

🤔 None of the numbers are exact. Are the threes I choose 50% "more" than the twos?

What might be a good comment? `// NOTE(cbandy): Just a guess` ?

jmckulk replied:
If none of them are exact then maybe it isn't worth a comment. Just assume that number doesn't matter, at least until the point where it does?

cbandy replied:
Added a comment to the couple `99`s in amended [`68658c5`](https://github.com/CrunchyData/postgres-operator/pull/2986/commits/68658c523e21a943287b7ef4a5149912e8c835cd).

cbandy commented:
Rebased and added a linter exception for the now-unused `setupTestEnv` and `teardownTestEnv` in [`68658c5`](https://github.com/CrunchyData/postgres-operator/pull/2986/commits/68658c523e21a943287b7ef4a5149912e8c835cd).
